resolwe_bio/processes/reads_processing/rnaseq_vc_preprocess.py

```diff
@@ -0,0 +1,411 @@
+            "-O",
+            intervals_path,
+        ]
+        return_code, stdout, stderr = Cmd["gatk"]["SplitIntervals"][
+            split_intervals_inputs
+        ] & TEE(retcode=None)
+        if return_code:
+            print(stdout, stderr)
+            self.error("SplitIntervals tool failed.")
+
+        # Prepare folder for SplitNCigarReads BAM files
+        output_bams = Path("output_bams")
+        output_bams.mkdir()
+
+        intervals = [path for path in intervals_path.glob("*.interval_list")]
+        return_codes = Parallel(n_jobs=n_jobs)(
+            run_split_ncigar_reads(
+                input_bam=marked_dups,
+                interval_path=interval_path,
+                ref_seq_path=inputs.ref_seq.output.fasta.path,
+                tmp=TMPDIR,
+                parent_dir=output_bams,
+            )
+            for interval_path in intervals
+        )
+        if any(return_codes):
+            self.error("GATK SplitNCigarReads tool failed.")
+
+        Path(marked_dups).unlink(missing_ok=True)
+        Path(marked_dups + ".bai").unlink(missing_ok=True)
+
+        input_lists = prepare_scattered_inputs(results_dir=output_bams, pattern="*.bam")
+        gather_bam_inputs = [
+            "-O",
+            splitncigar,
+            input_lists,
+            "--TMP_DIR",
+            TMPDIR,
+        ]
+
+        return_code, stdout, stderr = Cmd["gatk"]["GatherBamFiles"][
+            gather_bam_inputs
+        ] & TEE(retcode=None)
+        if return_code:
+            print(stdout, stderr)
+            self.error("GatherBamFiles tool failed.")
+
+        # Delete all files produces by SplitNCigarReads
+        for bam_file in output_bams.glob("*"):
+            Path(bam_file).unlink(missing_ok=True)
+        Path(output_bams).rmdir()
+
+        arg_rg = [
+            "--java-options",
+            f"-XX:ParallelGCThreads={gc_threads} -Xmx{inputs.advanced.max_heap_size}g",
+            "--INPUT",
+            splitncigar,
+            "--VALIDATION_STRINGENCY",
+            "STRICT",
+            "--OUTPUT",
+            read_groups,
+            "--TMP_DIR",
+            TMPDIR,
+        ]
+
+        present_tags = []
+        for x in inputs.read_group.split(";"):
+            split_tag = x.split("=")
+            arg_rg.extend(split_tag)
+            present_tags.append(split_tag[0])
+
+        # Make sure all arguments to read_group are valid.
+        all_tags = {
+            "-LB",
+            "-PL",
+            "-PU",
+            "-SM",
+            "-CN",
+            "-DS",
+            "-DT",
+            "-FO",
+            "-ID",
+            "-KS",
+            "-PG",
+            "-PI",
+            "-PM",
+            "-SO",
+        }
+        present_tag_set = set(present_tags)
+        check_all_tags = present_tag_set.issubset(all_tags)
+        if not check_all_tags:
+            self.error("One or more read_group argument(s) improperly formatted.")
```

Comment:
Also, since this block of code servers as a process input validation, can we put it in the beginning, so that the processing doesn't fail only after a lot of compute has been already done?